File: perl/csv2files.pl
#!/usr/bin/perl
#
use strict;
use warnings;

our $numFile = "theNumber.txt";

#my $descFile = "../dataVanJaak/boards.csv";
#my $descFile = "../dataVanJaak/ORG_IBMpc.csv";
#my $descFile = "../dataVanJaak/portables.csv";
my $descFile = "../dataVanJaak/ideHarddisks.csv";
#my $location = "JaakSite/boards/";
#my $location = "JaakSite/IBM/";
#my $location = "JaakSite/portables/";
my $location = "JaakSite/ideHarddisks/";


unless (-r $descFile && -e $descFile) { die "$0: $descFile not found. aborting."; }

open IFH, "<$descFile"  or die "$0: error opening \'$descFile\'";

my $line;
my $lineCount = 0;
my @propNames;
my @propValues;
my %props;

while ($line = <IFH>) {
    if ($lineCount == 0) {
        @propNames = &processNames($line);
    } else {
        @propValues = &processValues($line);
        my $propCount = @propNames;
        print " propCount = $propCount\n";
        for (my $i = 0; $i < $propCount; $i++) {
            if ($propNames[$i] eq "id") {
                $props{"id"} = &getNextNumber();
            } else {
                if (defined($propValues[$i]) && $propValues[$i] ne "") {
                    my $name = &sanitizeString($propNames[$i]);
                    unless ($name =~ m/^\s*$/) {
                        $props{$name} = $propValues[$i];
                    }
                }
            }
        }
        my $fileName = $location . &sanitizeString(&createFileName(%props));
        print "fileName: $fileName\n";
        open OFH, ">$fileName"  or die "$0: error opening \'$fileName\'";
        foreach my $key (keys %props) { print OFH "$key = $props{$key}\n"; }
        close OFH;
    }
    print "\n";
    $lineCount++;
}

sub processNames() {
    my ($line) = @_;
    my @names;
    chomp $line;
#    $line ~= 
    print "n: " . $line . "\n";
    (@names) = split(";", $line);
    return @names;
}

sub processValues() {
    my ($line) = @_;
    my @props;
    chomp $line;
    print "   " . $line . "\n";
    (@props) = split(";", $line);
    return @props;
}

sub createFileName($$) {
    (%props) = @_;
    my $name = $props{'fabrikant'} . "_" . $props{'model'} . "_" . $props{'id'} . ".properties";
    $name =~ s/\s+/_/g;
    return $name;
}

sub getNextNumber() {
    unless (-r $numFile && -e $numFile) { die "$0: PANIC: $numFile not found. aborting."; }
    open my $ifh, '<', $numFile or die "Can't open file $!";
    my $file_content = do { local $/; <$ifh> };
    close $ifh;

    chomp $file_content;
#    print $file_content . "\n";
    my $number = hex($file_content);
    $number++;
    $file_content = sprintf("%06X", $number);
#    print $file_content . "\n";

    unlink $numFile;
    open my $ofh, '>', $numFile or die "Can't open file $!";
    print $ofh $file_content . "\n";
    close $ofh;

    return $file_content
}

sub sanitizeString() {
    my ($name) = @_;
    $name =~ s/\s+/-/g;
    $name =~ s/[\|\/:;=\$\!\@\#%\^\&\*\(\)\[\]]/_/g;
    return $name;
}

File: python/csv2files.py
#!/usr/bin/python
#

import os
import re
import datetime
import sys

def getNextNumber():
    numberFile = "theNumber.txt"
    
    file = open(numberFile, 'r')
    count = int(file.readline().rstrip(), 16)
    file.close()
    
    count = count + 1
    
    file = open(numberFile, 'w')
    file.write(hex(count))
    file.close()
    return (hex(count))[2:]


def createObjectFile(dirBase, objectProps):
    id = getNextNumber()
    fileName = "_".join([objectProps["fabrikant"], objectProps["model"]])
    if ("name" in objectProps):
        "_".join([fileName, objectProps["name"]])
    fileName = dirBase + "/" + sanitizeString(fileName) + "_" + id + ".properties"

    print ("fileName: " + fileName)
    file = open(fileName, 'w')
    print ("id = " + id)
    file.write("id = " + id + "\n")
    for propKey in objectProps.keys():
        if (objectProps[propKey]):
            print (propKey + " = " + objectProps[propKey])
            file.write(propKey + " = " + objectProps[propKey] + "\n")
    file.close()
            
def getFileNameBase(path):
    return os.path.splitext(os.path.basename(path))[0]
    
def sanitizeString(name):
    #[\|\/:;=\$\!\@\#%\^\&\*\(\)\[\]]
    return re.sub("[\|\/:;=\$\!\@\#%\^\&\*\(\)\[\]\s]", "_", name)

################################################################################

csvFile = ""

targetDirBase = "root/inventaris"

if len(sys.argv) > 1:
    csvFile = sys.argv[1]
    
if (not csvFile):
    print ("No csv file specified. Exiting.")
    exit(1)
    
print ("source file: " + csvFile)

targetDir = targetDirBase + "/" + getFileNameBase(csvFile)

print ("target directory: " + targetDir)
    
file = open(csvFile, 'r')
lines = file.readlines()
file.close()

lineCount = 0
headerCount = 0
objectProps = {}
for line in lines:
    lineCount += 1
    line = line.rstrip()
    print (str(lineCount) + "  " + line.rstrip())
    if (lineCount == 1):
        headers = line.split(";")
        headerCount = len(headers)
    else:
        if (line.strip()):
            objectProps = {}
            fields = line.split(";")
            for i in range(headerCount):
                print (headers[i] + " = " + fields[i])
                objectProps[headers[i].strip()] = fields[i].strip()

        createObjectFile(targetDir, objectProps)
